src/Views/VPNPage.vala

```diff
@@ -274,50 +218,123 @@ public class Network.VPNPage : Network.Widgets.Page {
+    private void update_active_connections () {
+        if (active_connections == null) {
+            active_connections = new Gee.ArrayList<NM.VpnConnection> ();
         }
 
-        return (VPNMenuItem)children.last ().data;
-    }
-
-    private void update_active_connection () {
-        active_connection = null;
+        active_connections.clear ();
 
         unowned NetworkManager network_manager = NetworkManager.get_default ();
         network_manager.client.get_active_connections ().foreach ((ac) => {
-            if (ac.get_vpn () && active_connection == null) {
-                active_connection = (NM.VpnConnection)ac;
-                active_connection.vpn_state_changed.connect (update);
+            if (ac.get_vpn ()) {
+                active_connections.add ((NM.VpnConnection) ac);
+                (ac as NM.VpnConnection).vpn_state_changed.connect (update);
             }
         });
     }
 
-    private void vpn_activate_cb (VPNMenuItem item) {
-        active_vpn_item = item;
-        foreach (var child in vpn_list.get_children ()) {
-            ((VPNMenuItem)child).hide_icons ();
-        }
-
-        update ();
+    private void connect_vpn_cb (VPNMenuItem item) {
+        update_active_connections ();
         unowned NetworkManager network_manager = NetworkManager.get_default ();
         network_manager.client.activate_connection_async.begin (item.connection, null, null, null, null);
+        update ();
     }
 
-    private void vpn_deactivate_cb () {
-        update_active_connection ();
-        if (active_connection == null) {
-            return;
+    private void disconnect_vpn_cb (VPNMenuItem item) {
+        update_active_connections ();
+        unowned NetworkManager network_manager = NetworkManager.get_default ();
+        foreach (var ac in active_connections) {
+            if (ac.get_connection () == item.connection) {
+                try {
+                    network_manager.client.deactivate_connection (ac);
+                } catch (Error e) {
+                    warning (e.message);
+                }
+            }
         }
-
         update ();
-        unowned NetworkManager network_manager = NetworkManager.get_default ();
+    }
+
+    private void edit_connections () {
+        var command = "nm-connection-editor --type=vpn -s";
+
+        var selected_row = vpn_list.get_selected_row () as VPNMenuItem;
+        if (selected_row != null) {
+            command = "nm-connection-editor --edit=" + selected_row.connection.get_uuid ();
+        }
+
         try {
-            network_manager.client.deactivate_connection (active_connection);
+            var appinfo = AppInfo.create_from_commandline (
+                command,
+                null,
+                AppInfoCreateFlags.NONE
+            );
+            appinfo.launch (null, null);
+        } catch (Error e) {
+            warning ("%s", e.message);
+        }
+    }
+
+    private bool create_connection () {
```

Comment:
This can be static.

Reply:
Removed this and initialized it in constructor.